### PR TITLE
FEATURE: include themes and components keywords to the admin sidebar

### PR DESCRIPTION
**Pull request by lis2**
Include themes and component keywords to make the filter more accurate.

https://github.com/discourse/discourse/assets/72780/4f4d4ff7-c2fe-4169-9f2b-f140e03c62c8

### CODE REVIEW

File: app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js [hunk 1]
```diff
@@ -269,13 +270,30 @@ export default class AdminSidebarPanel extends BaseCustomSidebarPanel {
...
       this.adminSidebarStateManager.setLinkKeywords(
         "admin_installed_plugins",
         installedPluginsLinkKeywords()
       );
     }
 
+    store.findAll("theme").then((themes) => {
+      this.adminSidebarStateManager.setLinkKeywords(
+        "admin_themes",
+        themes.content.filter((theme) => !theme.component).mapBy("name")
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js:289]: I think you can use `rejectBy("component")` here.

File: app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js [hunk 2]
```diff
@@ -269,13 +270,30 @@ export default class AdminSidebarPanel extends BaseCustomSidebarPanel {
...
+      );
+      this.adminSidebarStateManager.setLinkKeywords(
+        "admin_components",
+        themes.content.filter((component) => component.component).mapBy("name")
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js:293]: I think you can use `filterBy("component")` here.

File: app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js [hunk 3]
```diff
@@ -269,13 +270,30 @@ export default class AdminSidebarPanel extends BaseCustomSidebarPanel {
     const navMap = savedConfig || ADMIN_NAV_MAP;
 
     if (!session.get("safe_mode")) {
-      navMap.findBy("name", "plugins").links.push(...pluginAdminRouteLinks());
+      const pluginLinks = navMap.findBy("name", "plugins").links;
+      pluginAdminRouteLinks().forEach((pluginLink) => {
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js:274]: Was this change necessary because we were doubling up with links at times or something?

File: app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js [hunk 4]
```diff
@@ -269,13 +270,30 @@ export default class AdminSidebarPanel extends BaseCustomSidebarPanel {
...
+        }
+      });
+
       this.adminSidebarStateManager.setLinkKeywords(
         "admin_installed_plugins",
         installedPluginsLinkKeywords()
       );
     }
 
+    store.findAll("theme").then((themes) => {
```
Comment (martin-brennan) [inline app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js:286]: I do worry about having an AJAX request here; this is why I added `visiblePlugins` to the preload store for admins:

https://github.com/discourse/discourse/blob/e1d9fd479f83191af24a10964a0e9915de6141f5/app/controllers/application_controller.rb#L674-L688

I think it might be best if we store a simplified array of themes & components preloaded for admin for this purpose, e.g. something like:

```ruby
Theme.include_relations.order(:name).to_a.map { |t| [t.name, t.component, t.enabled] }
```

You are right that eventually we want some serialized search data index for admin with all routes and keywords for the admin-wide search down the road -- I think also we would preload this, then we would be able to get rid of `visiblePlugins` and whatever this above preload store is called (`installedThemes`?)

**Review by martin-brennan — COMMENTED** [review]
Another consideration we need to have here is duplicate theme / component names, we need to dedupe when setting the link keywords in the sidebar:



We also may want to exclude the Default theme? Not sure. Might be some more special cases to think of e.g. if you go to https://meta.discourse.org/admin/customize/themes and see the tons of themes & components we have installed

File: app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js (hunk 3 above)
Comment (lis2) [reply app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js:274]: Yeah, that ajax call recalculates the cached function and could not prevent it, so this code must be temporarily idempotent

File: app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js (hunk 4 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/lib/sidebar/admin-sidebar.js:286]: We talked about this -- it's fine for now especially since it's just adding keywords and not blocking adding links. When we do the proper admin search we need to rethink all this search data anwyay.

**Comment by martin-brennan** [comment]
@lis2 did the `filterBy`/`rejectBy` stuff not work? Also need to address the duplicate names + default theme comments.

**Review by martin-brennan — APPROVED** [review]
